src/coreclr/jit/redundantbranchopts.cpp

```diff
@@ -1027,35 +1042,62 @@ bool Compiler::optRedundantRelop(BasicBlock* const block)
 
         JITDUMP(" -- prev tree is viable candidate for relop fwd sub!\n");
         candidateTree = prevTreeRHS;
-        reverse       = domIsRevRelop;
+        candidateStmt = prevStmt;
     }
 
     if (candidateTree == nullptr)
     {
         return false;
     }
 
-    // Looks good -- we are going to forward-sub candidateTree
-    //
-    GenTree* const substituteTree = gtCloneExpr(candidateTree);
+    GenTree* substituteTree = nullptr;
+    bool     usedCopy       = false;
 
-    // If we need the reverse compare, make it so
+    if (candidateStmt->GetNextStmt() == stmt)
+    {
+        // We are going forward-sub candidateTree
+        //
+        substituteTree = candidateTree;
+    }
+    else
+    {
+        // We going to forward-sub a copy of candidateTree
+        //
+        assert(!sideEffect);
+        substituteTree = gtCloneExpr(candidateTree);
+        usedCopy       = true;
+    }
+
+    // If we need the reverse compare, make it so.
+    // We also need to set a proper VN.
     //
-    if (reverse)
+    if ((vnRelationMatch == ValueNumStore::VN_RELATION_KIND::VRK_Reverse) ||
+        (vnRelationMatch == ValueNumStore::VN_RELATION_KIND::VRK_SwapReverse))
     {
+        // Copy the vn info as it will be trashed when we change the oper.
+        //
+        ValueNumPair origVNP = substituteTree->gtVNPair;
+
+        // Update the tree. Note we don't actually swap operands...?
+        //
         substituteTree->SetOper(GenTree::ReverseRelop(substituteTree->OperGet()));
```

Comment:
Is the comment incomplete?

Also, we can use `substitudeTree->SetOper(..., ValueNumberUpdate::PRESERVE_VN);` here instead.